// @summonergg/league-inhouse-api/src/models/game/translate/en_US.ts
import Discord from "discord.js";

export const en_US: Translater = (msg: DiscordMessage) => {
  return async (id, args?) => {
    var ranked = (args?.game?.ranked) ? "Ranked" : "Unranked";
    var res: Record<number, string> = {
      100: "Lobby has been canceled",

      200: "The max amount of players has been reached for this lobby",
      201: `${args?.player} was not found in the lobby`,
      202: `Please fill the lobby before you start`,
      203: `Not everyone in the lobby is ready`,

      300: `SummonerGG - ${ranked} - ${args?.game?.map}${args?.game?.draft}`,

      350: `<@${args?.user?.id}> Please select which pick order you want by typing \`fp\` or \`sp\``
    }

    if (id >= 350) { 
      args?.channel.send(res[id]).then((data: Discord.Message) => {
        setTimeout(() => {
          if (data.deletable) data.delete();
        }, 5000)
      })
    } else if (id === 300) {
      var message = new Discord.MessageEmbed({
        type: "rich",
        title: res[300],
        fields: []
      })
      if (args?.teams?.one && args?.teams?.two) {
        message.fields.push({
          name: "Team 1",
          value: args?.teams?.one,
          inline: true
        })
        message.fields.push({
          name: "Team 2",
          value: args?.teams?.two,
          inline: true
        })
      }
      if (args?.lobby.length != 0) {
        message.fields.push({
          name: "Lobby",
          value: args?.lobby,
          inline: false
        })
      }
      if (!args?.game?.message) {
        msg.channel.send(message).then(data => {
          args.game.message = data.id;
          args.game.save();
        })
      } else {
        try {
          var channel = await msg.channel.messages.fetch(args?.game?.message);
          channel.edit(message);
        } catch(err) { console.error(err) }
      }
    } else if (id >= 200) {
      msg.reply(res[id]).then((data: Discord.Message) => {
        setTimeout(() => {
          if (data.deletable) data.delete();
        }, 5000)
      })
    }
  }
}
