// @summonergg/league-inhouse-api/dist/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.client = exports.DiscordAPI = exports.User = exports.Token = exports.Guild = exports.Game = void 0;
const tslib_1 = require("tslib");
const dotenv_1 = tslib_1.__importDefault(require("dotenv"));
dotenv_1.default.config();
const axios_1 = tslib_1.__importDefault(require("axios"));
var game_1 = require("./models/game");
Object.defineProperty(exports, "Game", { enumerable: true, get: function () { return game_1.Game; } });
var guild_1 = require("./models/guild");
Object.defineProperty(exports, "Guild", { enumerable: true, get: function () { return guild_1.Guild; } });
var token_1 = require("./models/token");
Object.defineProperty(exports, "Token", { enumerable: true, get: function () { return token_1.Token; } });
var user_1 = require("./models/user");
Object.defineProperty(exports, "User", { enumerable: true, get: function () { return user_1.User; } });
var discord_1 = require("./models/discord");
Object.defineProperty(exports, "DiscordAPI", { enumerable: true, get: function () { return discord_1.DiscordAPI; } });
const IP = process.env.LEAGUE_INHOUSE_CLIENT_IP;
const PORT = process.env.LEAGUE_INHOUSE_CLIENT_PORT;
exports.client = axios_1.default.create({
    baseURL: `http://${IP}:${PORT}/api`
});
