// @summonergg/riot-api/src/modules/limiter.ts
import Bottleneck from "bottleneck";
import axios from "axios";
import { endpoints } from "./static";

export const bottleneck = () => {
  const RIOT_ID = process.env.RIOT_ID;
  const RIOT_RATE = process.env.RIOT_RATE;
  const RIOT_HOST = process.env.RIOT_HOST;
  const RIOT_PORT = process.env.RIOT_PORT;
  return new Bottleneck({
    datastore: "redis",
    clearDatastore: false,
    clientOptions: {
      host: RIOT_HOST ?? "127.0.0.1",
      port: RIOT_PORT ?? 6379
    },
    minTime: (1000/(+RIOT_RATE)) ?? (1000/.83),
    id: RIOT_ID ?? "riot-api"
  })
}

export const api: apiOption = async (url, options) => {
  const limiter = bottleneck();
  const RIOT_API_KEY = process.env.RIOT_API_KEY;
  const RIOT_REGION = process.env.RIOT_REGION;
  const RIOT_ID = process.env.RIOT_ID;

  var region = options.region ?? RIOT_REGION;
  var params = {
    params: {
      api_key: options.api_token ?? RIOT_API_KEY,
      ...options
    }
  }
  limiter.id = `${RIOT_ID}-${region}`;
  var res = await limiter.schedule(() => axios.get(`https://${endpoints[region].domain}${url}`, params));
  return res;
}
